pkg/deploy/spindeploy/transformer/patches.go

```diff
@@ -0,0 +1,204 @@
+	if err != nil {
+		return nil, err
+	}
+	c := &creator{}
+	gvk := obj.GetObjectKind().GroupVersionKind()
+	for i, patch := range kustomization.Patches {
+		originalObjJS, err = getPatchedJSON(types.MergePatchType, originalObjJS, []byte(patch), gvk, c)
+		if err != nil {
+			return nil, fmt.Errorf("unable to apply json patch for %s at index %d: %v", spinsvc, i, err)
+		}
+	}
+	if kustomization.PatchesJson6902 != "" {
+		originalObjJS, err = getPatchedJSON(types.JSONPatchType, originalObjJS, []byte(kustomization.PatchesJson6902), gvk, c)
+		if err != nil {
+			return nil, fmt.Errorf("unable to apply json6902 patch for %s: %v", spinsvc, err)
+		}
+	}
+	for i, patch := range kustomization.PatchesStrategicMerge {
+		originalObjJS, err = getPatchedJSON(types.StrategicMergePatchType, originalObjJS, []byte(patch), gvk, c)
+		if err != nil {
+			return nil, fmt.Errorf("unable to apply strategic merge patch for %s at index %d: %v", spinsvc, i, err)
+		}
+	}
+	return originalObjJS, nil
+}
+
+func (p *patchTransformer) asDeployment(data []byte) (*appsv1.Deployment, error) {
+	dser := scheme.Codecs.UniversalDecoder()
+	dep := &appsv1.Deployment{}
+	obj, _, err := dser.Decode(data, &schema.GroupVersionKind{
+		Group:   "apps",
+		Version: "v1",
+		Kind:    "Deployment",
+	}, dep)
+	if err != nil {
+		return nil, err
+	}
+	if dep, ok := obj.(*appsv1.Deployment); ok {
+		return dep, nil
+	}
+	return nil, errors.New("not parsed as a deployment, patching failed")
+}
+
+func (p *patchTransformer) asService(data []byte) (*v1.Service, error) {
+	dser := scheme.Codecs.UniversalDecoder()
+	svc := &v1.Service{}
+	obj, _, err := dser.Decode(data, &schema.GroupVersionKind{
+		Group:   "apps",
+		Version: "v1",
+		Kind:    "Deployment",
+	}, svc)
+	if err != nil {
+		return nil, err
+	}
+	if svc, ok := obj.(*v1.Service); ok {
+		return svc, nil
+	}
+	return nil, errors.New("not parsed as a service, patching failed")
+}
+
+type creator struct{}
+
+func (c *creator) New(gvk schema.GroupVersionKind) (runtime.Object, error) {
+	if gvk.Kind == "Deployment" {
+		return &appsv1.Deployment{}, nil
+	}
+	if gvk.Kind == "Service" {
+		return &v1.Service{}, nil
+	}
+	return nil, fmt.Errorf("unknown gvk to patch %s", gvk.String())
+}
+
+func getPatchedJSON(patchType types.PatchType, originalJS, patchJS []byte, gvk schema.GroupVersionKind, creater runtime.ObjectCreater) ([]byte, error) {
+	patchJS, err := yaml.ToJSON(patchJS)
+	if err != nil {
+		return nil, err
+	}
+
+	switch patchType {
+	case types.JSONPatchType:
+		patchObj, err := jsonpatch.DecodePatch(patchJS)
+		if err != nil {
+			return nil, err
+		}
+		bytes, err := patchObj.Apply(originalJS)
+		// TODO: This is pretty hacky, we need a better structured error from the json-patch
+		if err != nil && strings.Contains(err.Error(), "doc is missing key") {
+			msg := err.Error()
+			ix := strings.Index(msg, "key:")
+			key := msg[ix+5:]
+			return bytes, fmt.Errorf("Object to be patched is missing field (%s)", key)
```

Comment:
It's probably good to wrap the original error to provide more context:
```
return bytes, fmt.Errorf("Object to be patched is missing field (%s). Original error: %w", key, err)
```
https://blog.golang.org/go1.13-errors (Wrapping errors with %w)

Reply:
This is how kustomize does it: error is swallowed because the underlying one doesn't make sense to the user. I'd rather keep it that way.